### ISSUE TITLE
Snapshot, summary, Tree entries: Need standardization and reduction of concepts.

### ISSUE THREAD

**Issue by vladsud** [microsoft/FluidFramework#4683]
Our File I/O types are very confusing - it's very hard to follow code on where we use certain types and why:

ISnapshotTree - used when loading from storage, could have name collisions between blobs & trees.
ITree / ITreeEntry - entries, could have name collisions; used when attaching DDS / data store, also used on legacy path
git.ITree - somewhat similar shape as above, mostly used in R11s, but leaks to various helper functions
ISummaryTree - summarization process, no collisions, supports symbolic reuse of content (through handle + path used as a reference)

We have a bunch of helper functions to convert from one type to another:
convertSnapshotTreeToSummaryTree: ISnapshotTree -> ISummaryTreeWithStats
convertToSummaryTree: ITree -> ISummarizeResult
convertSummaryTreeToITree : ISummaryTree -> ITree
buildSnapshotTree: ITree -> ISnapshotTree 

While we can't completely remove all these types from runtime (as some are tied to file format), I believe we could reduce it to a minimum, by always converting any time to summary type and all the code working only with summaries.
This would substantially reduce mental workload when working with all these layers.

This work can be done incrementally, but adding more converter functions to support any to any conversions, and thus allowing us to address layer by layer, keeping system functional during this process.

Related item: https://github.com/microsoft/FluidFramework/pull/4682 (Cleanup of summary related interfaces)

**Comment by vladsud** [comment]
@arinwt - would love you taking a look and providing feedback,
@curtisman, @danielroney - I'd love to see some mechanism to get it scheduled and not sit in "Next" bucket forever.

**Comment by vladsud** [comment]
tagging @agarwal-navin

**Comment by agarwal-navin** [comment]
Yes, I agree that we need to standardize summaries and snapshots. In my GC work, I came across these and got confused as well. More than that, I had to add the GC data across these formats and make it compatible and make sure there are no bugs. Standardizing this will help making changes across this space easier and less error prone.

IMO, there are 2 major chunks of work that need to be cleaned up -
1. DDSs should start using generating ISummaryTree in summarizer.
2.  Container should stop exposing snapshot and expose summarize instead.

**Comment by vladsud** [comment]
@scarlettjlee, since you have been poking in this area (indirectly), I hope you do not mind if I assign this item to you.
I've just added an item on your plate (https://github.com/microsoft/FluidFramework/issues/8652) to this Epic.

We can chat more about details, priority, and what else needs to be included into this Epic.

**Comment by agarwal-navin** [comment]
@scarlettjlee I added https://github.com/microsoft/FluidFramework/issues/4728 to this list. Please feel free to remove  it from here if it seems not relevant to this epic.

**Comment by scarlettjlee** [comment]
@vladsud & @agarwal-navin, could you review the updated [sub-]issues list in the original comment up top?

**Comment by vladsud** [comment]
Looks reasonable to me (though it does feel similar to what it used to be when I opened it, maybe minus sub-items).
I'd say the most interesting question is the ordering, and what would we do if we have only 1 month (or whatever amount of time we have). I'd use simple rule to make these decisions - if we do not do some work, does it become more expensive over time (because more usage cripples in), or does it make other developers spend more time to understand more concepts and deal with them? If the answer is yes, that's candidate for higher priority.

The opposite of that would be something like commits on ISnapshotTree. It's not used, and will not be used in runtime, so it's the cleanup that we would love to do, but it's not becoming more expensive over time.

I'd also optimize towards consistency. We should consistently use "summary" anywhere we mean upload, and snapshot anywhere we mean download format. That should be consistent in comments, names of functions, interfaces.

**Comment by scarlettjlee** [comment]
@vladsud, I reorganized the task list to try to categorize them. A bunch of the desired work has already happened since this item was first opened. I'd focus any concerted effort toward getting rid of ITree, especially #9227. And then unifying snapshot and summary (which I haven't thought much about yet). The cleanup tasks can be done whenever and aren't causing much trouble (maybe some "good first issue"s that we could do opportunistically). There's some other minor misusage of terms in the codebase that I don't think are serious but would be nice to fix, (like "snapshot" in matrix DDS to get DDS content that is not a snapshot type).

**Comment by ssimic2** [comment]
Related: https://github.com/microsoft/FluidFramework/issues/8979

**Comment by vladsud** [comment]
Thanks for clarification!
There is no way to unify snapshot & summary interface, as they have different qualities (snapshot does not have blob content, but otherwise self-contained, i.e., can't refer to content from other snapshots. while summaries always include blob content, unless they refer to blobs in storage / previous snapshot).

If you feel like new tasks are very clear and Ok for new hires, please mark them as first issue.
